mlir/include/mlir/IR/BuiltinTypes.h

```diff
@@ -191,6 +191,60 @@ class BaseMemRefType : public Type, public ShapedType::Trait<BaseMemRefType> {
 
 namespace mlir {
 
+//===----------------------------------------------------------------------===//
+// CopyOnWriteArrayRef<T>
+//===----------------------------------------------------------------------===//
+
+// A wrapper around an ArrayRef<T> that copies to a SmallVector<T> on
+// modification. This is for use in the mlir::<Type>::Builders.
+template <typename T>
+class CopyOnWriteArrayRef {
```

Comment:
Seems like a nice abstraction! 
Should this go in a more general header in `mlir/include/mlir/Support/` ?

Reply:
Could do! It is rather bare-bones, so any future users may need to further extend it :)